osu.Server.Spectator/Hubs/StatefulUserHub.cs

```diff
@@ -2,94 +2,107 @@
         {
-            this.Cache = cache;
         }
 
+        protected static KeyValuePair<long, TUserState?>[] GetAllStates() => ACTIVE_STATES.GetAllEntities();
+
         /// <summary>
         /// The osu! user id for the currently processing context.
         /// </summary>
         protected int CurrentContextUserId => int.Parse(Context.UserIdentifier);
 
-        public override Task OnConnectedAsync()
+        public override async Task OnConnectedAsync()
         {
             Console.WriteLine($"User {CurrentContextUserId} connected!");
 
-            return base.OnConnectedAsync();
-        }
+            // if a previous connection is still present for the current user, we need to clean it up.
+            await cleanUpState(false);
 
-        /// <summary>
-        /// Called when a user disconnected, providing their last state.
-        /// </summary>
-        /// <param name="exception">A potential error which caused the disconnection.</param>
-        /// <param name="state">The last user state. May be null. This is automatically cleared on disconnection.</param>
-        protected virtual Task OnDisconnectedAsync(Exception exception, TUserState? state) => Task.CompletedTask;
+            await base.OnConnectedAsync();
+        }
 
         public sealed override async Task OnDisconnectedAsync(Exception exception)
         {
             Console.WriteLine($"User {CurrentContextUserId} disconnected!");
 
-            var state = await GetLocalUserState();
-
-            await OnDisconnectedAsync(exception, state);
-
-            // clean up user on disconnection
-            if (state != null) await RemoveLocalUserState();
-
-            await base.OnDisconnectedAsync(exception);
+            await cleanUpState(true);
         }
 
-        protected async Task UpdateLocalUserState(TUserState state)
+        private async Task cleanUpState(bool isDisconnect)
         {
-            ACTIVE_STATES.TryRemove(CurrentContextUserId, out var _);
-            ACTIVE_STATES.TryAdd(CurrentContextUserId, state);
-
-            await Cache.SetStringAsync(GetStateId(CurrentContextUserId), JsonConvert.SerializeObject(state));
+            ItemUsage<TUserState>? usage;
+
+            try
+            {
+                usage = await ACTIVE_STATES.GetForUse(CurrentContextUserId);
+            }
+            catch (ArgumentException)
+            {
+                // no state to clean up.
+                return;
+            }
+
+            using (usage)
+            {
+                if (usage.Item != null)
+                {
+                    bool isOurState = usage.Item.ConnectionId != Context.ConnectionId;
+
+                    if (isDisconnect && !isOurState)
+                        // not our state, owned by a different connection.
+                        return;
+                }
+
+                if (usage.Item != null)
+                    await CleanUpState(usage.Item);
```

Comment:
Any reason not to merge this into the conditional above? This is all under a lock, so there should be no volatile changes, should there?

Reply:
No reason, just an oversight